### ISSUE TITLE
No secure MAC methods available

### ISSUE THREAD

**Issue by richardsimko** [jakob/Postico#713]
## What did you do?
When using the "Connect via SSH" feature the MACs available do not match those recommended for a secure sshd config. I'm guessing this is due to using an older version of openssh in Postico. If I connect using the system ssh it works as expected and the `chacha20-poly1305@openssh.com` cipher is used.

## What did you expect to happen?
Connect using secure MAC methods. The server is configured to allow the following ciphers and MACs:
```
Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,umac-128-etm@openssh.com
```

## What actually happened?

Postico can not connect. Server logs the following:
```
Jun  2 10:45:16 do-fra1-postgres sshd[9599]: Unable to negotiate with xxxx port 40326: no matching MAC found. Their offer: hmac-sha2-256,hmac-sha2-512,hmac-sha1,hmac-sha1-96,hmac-md5,hmac-md5-96,hmac-ripemd160,hmac-ripemd160@openssh.com [preauth]
```

The least bad of these (`hmac-sha2-512`) is still recommended by tools such as [ssh-audit](https://github.com/arthepsy/ssh-audit) to be removed.

## What software versions are you using?

Postico version:  
1.5.13

macOS version:  
10.15.4

PostgreSQL version:  
12-alpine

**Comment by jakob** [comment]
Postico 2 uses libssh2 for SSH tunnels: https://www.libssh2.org/

We're not crypto experts, so it's unlikely that we are going to work on adding newer ciphers to libssh2. We can only support what libssh2 supports.

(The list of supported crypto on the homepage is not yet updated for the latest version, as EC and Ed25519 keys are now also supported)

However, I'm not quite sure what the issue is. There are a few ciphers and hmacs that both your server and Postico should support.

Finally, I'm aware that SHA-2 isn't the most modern MAC, but AFAIK it's not broken? Does ssh-audit say why you shouldn't use it?

Our goal with Postico is to support most commonly used SSH server configurations. Unfortunately we can't support all configurations, so if your require specific algorithms my recommendation is to use OpenSSH to manually open SSH tunnels. This would probably be the safer choice anyway, since a security vulnerability in Postico's SSH code is at least an order of magnitude more likely than a practical attack  on any of the legacy algorithms that libssh2 supports.

**Comment by richardsimko** [comment]
Thanks for the info, that's fair enough. libssh2 seems to have relatively poor support for the more secure ciphers according to https://ssh-comparison.quendi.de/comparison/cipher.html

> However, I'm not quite sure what the issue is. There are a few ciphers and hmacs that both your server and Postico should support.
My bad, I updated the sshd_config file while troubleshooting and pasted the updated version (Where I added a compatible MAC). I've updated the issue.

> Finally, I'm aware that SHA-2 isn't the most modern MAC, but AFAIK it's not broken? Does ssh-audit say why you shouldn't use it?

The main reason for recommending against sha-2 is not that it's broken but that it uses the encrypt-and-MAC mode. I'm not a security expert myself either so I will go with the explanation and recommendation of those who are. This [stackoverflow thread](https://crypto.stackexchange.com/a/205) has more details on the difference.

My general approach to security is that if avoiding an attack vector is close to no effort (As it is with disabling these MACs and ciphers for sshd) it's worth doing, however unlikely the attack vector is.

Anyhow, thanks for the info, I will stick to using manual SSH tunnels for now :)
